dask/dataframe/io/parquet/arrow.py

```diff
@@ -265,149 +91,540 @@ def _index_in_schema(index, schema):
-            (
-                index_names,
-                column_names,
-                storage_name_mapping,
-                column_index_names,
-            ) = _parse_pandas_metadata(pandas_metadata)
-            if categories is None:
-                categories = []
-                for col in pandas_metadata["columns"]:
-                    if (col["pandas_type"] == "categorical") and (
-                        col["name"] not in categories
-                    ):
-                        categories.append(col["name"])
-        else:
-            index_names = []
-            column_names = schema.names
-            storage_name_mapping = {k: k for k in column_names}
-            column_index_names = [None]
-
-        if index is None and index_names:
-            index = index_names
-
-        if set(column_names).intersection(partitions):
-            raise ValueError(
-                "partition(s) should not exist in columns.\n"
-                "categories: {} | partitions: {}".format(column_names, partitions)
+    dtypes = _get_pyarrow_dtypes(schema, categories)
+    dtypes = {storage_name_mapping.get(k, k): v for k, v in dtypes.items()}
+
+    index_cols = index or ()
+    meta = _meta_from_dtypes(all_columns, dtypes, index_cols, column_index_names)
+    meta = clear_known_categories(meta, cols=categories)
+
+    if partition_obj:
+        for partition in partition_obj:
+            if isinstance(index, list) and partition.name == index[0]:
+                meta.index = pd.CategoricalIndex(
+                    categories=partition.keys, name=index[0]
+                )
+            elif partition.name == meta.index.name:
+                meta.index = pd.CategoricalIndex(
+                    categories=partition.keys, name=meta.index.name
+                )
+            elif partition.name in meta.columns:
+                meta[partition.name] = pd.Series(
+                    pd.Categorical(categories=partition.keys, values=[]),
+                    index=meta.index,
+                )
+
+    return meta, index_cols, categories, index
+
+
+def _aggregate_stats(
+    file_path, file_row_group_stats, file_row_group_column_stats, stat_col_indices
+):
+    """ Utility to aggregate the statistics for N row-groups
+        into a single dictionary.
+    """
+    if len(file_row_group_stats) < 1:
+        raise ValueError("Empty statistics.")
+    elif len(file_row_group_column_stats) == 0:
+        assert len(file_row_group_stats) == 1
+        return file_row_group_stats[0]
+    else:
+        df_rgs = pd.DataFrame(file_row_group_stats)
+        s = {
+            "file_path_0": file_path,
+            "num-rows": df_rgs["num-rows"].sum(),
+            "total_byte_size": df_rgs["total_byte_size"].sum(),
+            "columns": [],
+        }
+        df_cols = pd.DataFrame(file_row_group_column_stats)
+        for ind, name in enumerate(stat_col_indices):
```

Comment:
So we still reformat for all columns here? There was a separate discussion not to bother doing this for non-index columns (until we figure out how to push down generalised filters)

Reply:
Also, *this* is where early-bail would happen for any column that is not a good order

Reply:
At this point (in `_aggregate_stats`) we already have the metadata and statistics for the subset of columns we "need" it for.  `stat_col_indices` only contains the name of these columns that we already collected statistics for (not all columns).  In `_process_metadata`, we would have bailed on statistics gathering (and avoided calling `_aggregate_stats` altogether) if we only gathering statistics for an unordered column/index.

Reply:
got it